File: frontend/src/types/device.ts
/** Types for devices, their status and alerts. */

/** Supported platforms. Mirrors `DeviceType` in the backend constants. */
export type DeviceType = 'ios' | 'android' | 'windows' | 'macos';

/**
 * Device reachability.
 *
 * `UNKNOWN` means the device reported a WiFi MAC other than the one it
 * registered with -- it is reachable but no longer part of this alert group,
 * so the UI must not offer to beep it.
 */
export type DeviceStatus = 'ONLINE' | 'OFFLINE' | 'UNKNOWN';

export type AlertStatus = 'SENT' | 'RECEIVED' | 'FAILED';

export interface Device {
  device_id: string;
  device_name: string | null;
  device_type: DeviceType;
  device_os_version: string | null;
  /** Null on platforms that do not report a battery. */
  battery_level: number | null;
  status: DeviceStatus;
  last_heartbeat: string | null;
  created_at: string;
  /**
   * True for a device that auto-registered without an account.
   *
   * A guest receives alerts like any other device but cannot send them. The
   * dashboard shows a "Guest" badge and disables its alert button.
   */
  is_guest: boolean;
}

export interface DeviceRegisterRequest {
  device_name: string;
  device_type: DeviceType;
  device_os_version?: string;
  push_token: string;
  wifi_mac: string;
  network_name?: string;
  /**
   * Stable across reinstalls and push-token rotations.
   *
   * The backend matches a re-registration on this rather than on the push
   * token, which the platform reissues on every reinstall -- that is what left
   * one phone showing up in the dashboard once per install it had ever had.
   */
  install_id?: string;
}

/**
 * Result of registering.
 *
 * `device_token` is present only for a guest registration -- it authorises
 * that device's heartbeat and nothing else. An owned device gets null and uses
 * its user JWT instead.
 */
export interface DeviceRegisterResponse {
  device_id: string;
  is_guest: boolean;
  device_token: string | null;
}

export interface HeartbeatRequest {
  battery_level: number | null;
  wifi_mac: string;
}

export interface SendAlertRequest {
  /** Empty targets every device on the sender network. */
  device_ids: string[];
}

export interface AlertDeliveryStatus {
  device_id: string;
  device_name: string | null;
  status: AlertStatus;
  error_code: string | null;
}

export interface SendAlertResponse {
  alert_id: string;
  delivery_status: AlertDeliveryStatus[];
}

export interface AlertLog {
  alert_id: string;
  target_devices: string[];
  status: AlertStatus;
  created_at: string;
}

/** Frame pushed over the status WebSocket. */
export interface DeviceStatusUpdate {
  device_id: string;
  status: DeviceStatus;
  battery: number | null;
  timestamp: string;
}

/**
 * Something seen on the WiFi that is not a registered device.
 *
 * Not a `Device` and deliberately a separate type: it has no push token, no
 * status and no battery, and can never be alerted. A discovered smart TV has
 * no app on it to ring. Keeping the two apart is what stops a Send alert
 * button appearing over a printer.
 */
export interface DiscoveredDevice {
  discovered_id: string;
  ip_address: string;
  device_name: string | null;
  device_type: string | null;
  /** MDNS learned a name; SWEEP only learned that something answered. */
  discovered_via: string;
  first_seen: string;
  last_seen: string;
}

/** Body of POST /devices/scan. */
export interface ScanSubmission {
  wifi_mac: string;
  devices: {
    ip_address: string;
    device_name: string | null;
    device_type: string | null;
    discovered_via: string;
  }[];
}

/** Reply to POST /devices/scan. */
export interface ScanSubmissionResponse {
  recorded: number;
}

File: backend/src/schemas/device.py
"""Request and response contracts for device endpoints."""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from src.utils.constants import (
    MAX_BATTERY_LEVEL,
    MIN_BATTERY_LEVEL,
    DeviceStatus,
    DeviceType,
)
from src.utils.validators import MAC_ADDRESS_PATTERN


def _validate_mac(value: str) -> str:
    """Reject a malformed MAC at the API edge.

    The WiFi MAC is the trust boundary for alert authorization, so a bad value
    is refused rather than normalised into something that might match a real
    network by accident.
    """
    if not MAC_ADDRESS_PATTERN.match(value):
        raise ValueError("must be a MAC address, e.g. 00:1A:2B:3C:4D:5E")
    return value


class DeviceRegisterRequest(BaseModel):
    """Body of POST /devices/register.

    Identical for owned and guest registrations. Which one happens is decided
    by whether the request carries a user token, not by anything in the body --
    a client cannot ask to be an owner.
    """

    device_name: str = Field(max_length=255)
    device_type: DeviceType
    device_os_version: str | None = Field(default=None, max_length=50)
    push_token: str = Field(max_length=500)
    wifi_mac: str
    network_name: str | None = Field(default=None, max_length=255)
    # Stable across reinstalls and token rotations, which the push token is
    # not. Optional: a client older than this field still registers through
    # the push-token path, it simply leaves a row behind when it reinstalls.
    install_id: str | None = Field(default=None, max_length=255)

    _check_mac = field_validator("wifi_mac")(_validate_mac)


class HeartbeatRequest(BaseModel):
    """Body of PUT /devices/{device_id}/heartbeat.

    Sent every HEARTBEAT_INTERVAL_SECONDS. wifi_mac is re-sent each time so the
    server can detect a device that has moved off the registered network.
    """

    battery_level: int | None = Field(default=None, ge=MIN_BATTERY_LEVEL, le=MAX_BATTERY_LEVEL)
    wifi_mac: str

    _check_mac = field_validator("wifi_mac")(_validate_mac)


class DeviceResponse(BaseModel):
    """Public view of a device. Never includes the push token."""

    model_config = {"from_attributes": True}

    device_id: uuid.UUID
    device_name: str | None
    device_type: DeviceType
    device_os_version: str | None
    battery_level: int | None
    status: DeviceStatus
    last_heartbeat: datetime | None
    created_at: datetime
    # Derived from Device.user_id being null. The dashboard shows a "Guest"
    # badge and disables the alert button when true.
    is_guest: bool


class DeviceRegisterResponse(BaseModel):
    """Result of POST /devices/register.

    ``device_token`` is returned only for a guest registration. It authorises
    that one device's heartbeat and nothing else -- an owned device uses its
    user's JWT instead and gets null here.
    """

    device_id: uuid.UUID
    is_guest: bool
    device_token: str | None = None
